carsus/io/output/base.py

```diff
@@ -240,188 +260,222 @@ def calculate_collisional_strength(row, temperatures,
 
-            except (KeyError, TypeError):
-                continue
+        if self.cmfgen_reader is not None:
+            cf_lines = self.cmfgen_reader.lines
+            cf_lines['ds_id'] = 5
+        else:
+            cf_lines = pd.DataFrame(columns=gf_lines.columns)
 
-            df = self.get_lvl_index2id(df, self.levels_all, ion)
-            df['source'] = 'gfall'
-            gf_list.append(df)
+        lines = pd.concat([gf_lines, ch_lines, cf_lines], sort=True)
+        lines = lines.reset_index()
+        lines = lines.rename(columns={'ion_charge': 'ion_number'})
+        lines['line_id'] = range(1, len(lines)+1)
 
-        ch_list = []
-        logger.info('Ingesting lines from Chianti.')
+        # Filter lines by priority
         for ion in self.chianti_ions:
+            mask = (lines['ds_id'] != 4) & (
+                        lines['atomic_number'] == ion[0]) & (
+                            lines['ion_number'] == ion[1])
+            lines = lines.drop(lines[mask].index)
+
+        for ion in self.cmfgen_ions:
+            mask = (lines['ds_id'] != 5) & (
+                        lines['atomic_number'] == ion[0]) & (
+                            lines['ion_number'] == ion[1])
+            lines = lines.drop(lines[mask].index)
+
+        lines = lines.set_index(['atomic_number', 'ion_number'])
+        lines = lines.sort_index()  # To supress warnings
+        ions = set(self.gfall_ions).union(set(self.chianti_ions))\
+                    .union((set(self.cmfgen_ions)))
+
+        logger.info('Matching lines and levels.')
+        lns_list = [ self.get_lvl_index2id(lines.loc[ion], self.levels_all)
+                        for ion in ions]
+        lines = pd.concat(lns_list, sort=True)
+        lines = lines.set_index('line_id').sort_index()
 
-            df = ch.lines.loc[ion]
-            df['line_id'] = range(start, len(df) + start)
-            start = len(df) + start
-
-            df = self.get_lvl_index2id(df, self.levels_all, ion)
-            df['source'] = 'chianti'
-            ch_list.append(df)
-
-        df_list = gf_list + ch_list
-        lines = pd.concat(df_list, sort=True)
         lines['loggf'] = lines['gf'].apply(np.log10)
-
-        lines.set_index('line_id', inplace=True)
         lines.drop(columns=['energy_upper', 'j_upper', 'energy_lower',
                             'j_lower', 'level_index_lower',
                             'level_index_upper'], inplace=True)
 
+        gfall_mask = lines['ds_id'] == 2
+        chianti_mask = lines['ds_id'] == 4
+
+        # TODO: manage units consistently across all readers (this 
+        # could break compatibility with SQL)
+        lines.loc[~chianti_mask, 'wavelength'] = lines.loc[
+            ~chianti_mask, 'wavelength'].apply(lambda x: x*u.nm)
+
+        lines.loc[chianti_mask, 'wavelength'] = lines.loc[
+            chianti_mask, 'wavelength'].apply(lambda x: x*u.AA)
+
+        lines['wavelength'] = lines['wavelength'].apply(lambda x: \
```

Comment:
convert beforehand.

Reply:
I think it's too sensitive to change this before merging. Unit conversions were made with a LOT of care, it can be improved for sure, in the future.

Reply:
We should simplify this. Just multiply by u.nm.to(u.AA) at the beginning to do the unit conversion. There is no need for a lambda function here.